Проектная работа по седьмому спринту

SomnyMosh opened the pull request.

LavraSource committed:
what

LavraSource committed:
what

LavraSource committed:
what

LavraSource committed:
Ошибка ввода исправлена

LavraSource committed:
Создан класс "Menu"

LavraSource committed:
Убраны лишние строки кода в "Search"

LavraSource committed:
Небольшие поправки

Anna-Malinina commented on src/main/kotlin/Class.kt:
```diff
@@ -0,0 +1,178 @@
+            println("Пожалуйста, выберите одну из предоставленных опций. Введите только цифру опции без других символов")
+            line = readln()
+        }
+        if (line == "1") {
```
Удобнее было бы использовать конструкцию when, и было бы чуть более читабельно

Anna-Malinina commented on src/main/kotlin/Class.kt:
```diff
@@ -0,0 +1,178 @@
+            println("5. Вернуться в исходную директорию")
+        }
+        line = readln()
+        while (line != "1" && line != "2" && line != "3" && line != "4" && line != "5") {
```
Можно было бы написать такое условие: line.toIntOrNull() !in 1..5, где строка пытается привестись к инту, и в случае неуспеха стала бы null. 1..5 Дает нам множество чисел от 1 до 5 включительно, а оператор !in говорит нам условие, что строка НЕ входит в этот интервал (за счет отрицания "1")

Anna-Malinina replied:
Также toIntOrNull можно было бы сделать чуть раньше, в самой переменной, чтобы и везде ниже работать с числом или null

Anna-Malinina commented on src/main/kotlin/Search.kt:
```diff
@@ -0,0 +1,153 @@
+            }
+        }
+        when (type) {
+            "Check" -> {
```
Типы было бы удобнее вынести в константы, это бы снизило вероятность ошибки от опечатки, а также если бы пришлось немного поменять значение, чтобы не искать по всему коду

Anna-Malinina commented on src/main/kotlin/Search.kt:
```diff
@@ -0,0 +1,153 @@
+            array2=result3.toCharArray()
+            result3 =""
+            for (i in array2.indices) {
+                if (i>3){
```
Чтобы просто так не итерироваться по индексам 0, 1, 2, можно было бы в цикле написать i in 3..array2.indices

Anna-Malinina replied:
А еще лучше было бы просто взять срез с 3 индекса по последний, используя метод slice

Anna-Malinina commented on src/main/kotlin/Class.kt:
```diff
@@ -0,0 +1,178 @@
+            println("Содержание архива '${currentFolder(filename1)}':")
+        }
+        search("NoMatter", filename1, true, "Search")
+        if (isNote) {
```
Здесь блоки if и else по сути дублируют друг друга, разница в передаваемых параметрах, которые проверяется и if, то есть isNote. Пр. : `if (search("NoMatter",filename1, true, "IfEmpty"))` и `if (search("NoMatter",filename1, false, "IfEmpty"))`. Это можно решить, передавая параметр isNote напрямую, что позволит объединить повторяющиеся места, а конкретику уже добавлять лямбдами/доп методами и тд

Anna-Malinina commented on src/main/kotlin/Class.kt:
```diff
@@ -0,0 +1,178 @@
+
+    private fun create(filename1: String, isNote: Boolean) {
+        var line: String
+        if (isNote) {
```
Аналогичная ситуация здесь, если присмотреться, то логика очень похожа в обоих блоках

```
println("Введите название архива")
            line = readln()
            if (line == "") {
                while (line == "") {
                    println("Название архива не может быть пустым, попробуйте снова")
                    line = readln()
                }
            } else if (search(line, filename1, false, "Check")) {
                while (search(line, filename1, false, "Check")|| line == "") {
                    println("Архив с данным названием уже существует. Попробуйте снова.")
                    line = readln()
                }
            }
```
```
println("Введите название заметки")
            title = readln()
            if (title == "") {
                while (title == "") {
                    println("Название заметки не может быть пустым, попробуйте снова")
                    title = readln()
                }
            } else if (search(title, filename1, true, "Check")) {
                while (search(title, filename1, true, "Check") || title == "") {
                    println("Заметка с данным названием уже существует. Попробуйте снова.")
                    title = readln()
                }
            }
```

Anna-Malinina commented on src/main/kotlin/Search.kt:
```diff
@@ -0,0 +1,153 @@
+        if ((result2+"\\").length == count){
+            check = false
+        }
+        if (result == "Main.kt"||result=="Search.kt"||result=="Class.kt"||result==""){
```
Нужно ли выполнять метод дальше или можно было бы уже в этом ифе вернуть false?

Anna-Malinina commented on src/main/kotlin/Search.kt:
```diff
@@ -0,0 +1,153 @@
+        }
+        var array2 = result.toCharArray()
+        var result3 = ""
+        for (i in array2.indices.reversed()) { // итерируемся по массиву в обратном направлении
```
Можно было бы упростить: https://stackoverflow.com/questions/61187813/find-a-file-in-kotlin

LavraSource committed:
Убрано дублирование в основном коде

LavraSource committed:
"Search" оптимизирован

LavraSource committed:
Убрано отображение ".txt" для заметок

LavraSource committed:
Ещё немного оптизирован "search"

LavraSource committed:
Предотвращение потенциальных опечаток при использовании функции "search"

LavraSource committed:
Без "жёлтых" проблем теперь

SomnyMosh committed:
Описания

Добавлены описания, так как названия не очень хорошо описывают предназначение enum констант

Anna-Malinina commented on src/main/kotlin/Class.kt:
```diff
@@ -16,144 +11,118 @@ class Menu {
+        var texxt: String
+        println("Введите название $altPrint")
+        title = readln()
+        if (title == "") {
```
В этом if скрыт баг, так как проверка на существующую директорию только в блоке else. Если я сначала введу пустой title, я попаду в блок if, затем я введу название, которое уже существует. Так как проверка на существование есть только в блоке else, то программа упадет с ошибкой FileAlreadyExistsException. Эти проверки стоит проводить в одном цикле, как в блоке else, только внутри разграничить сообщения о пустом архиве, или об существующем

Anna-Malinina commented on src/main/kotlin/Class.kt:
```diff
@@ -16,144 +11,118 @@ class Menu {
+        if (search("NoMatter", filename1, isNote, SearchType.IF_EMPTY)) {
+            println("Пожалуйста, выберите одну из предоставленных опций. Впишите название $altPrint")
+            line = readln()
+            if (search(line, filename1, isNote, SearchType.CHECK)) {
```
Здесь тоже идет повторение логики. Как ее можно было бы убрать: 
1) Убрать этот if 
2) `while (!search(line, filename1, isNote, SearchType.CHECK))`  вынести вместо него перед 
```
if (isNote) {
    readTxt(line, filename1)
} else {
    screen(selectArchive(line, filename1))
}
```
Таким образом, этот цикл заменил бы проверку в этом if. Ведь если он пропустился, значит, условие и так выполнилось. Если нет, то он будет работать до ввода корректного названия

SomnyMosh committed:
Баг исправлен
